### DISCUSSION TITLE
Nemo file tagging

### DISCUSSION THREAD

**Discussion by Qronikarz** [linuxmint/discussions#184] [Ideas]
There were few requests for this on Nemo repository, but they just got closed and never went anywhere.

Basically the ability to tag files and then search using the tags. MacOS is said to have fantastic tagging support, on Windows there was the option, but I never used it. And KDE Plasma is one of the few Linux DEs with tagging support.

There was a work in the Freedesktop specs in the past, but it was active long time ago. https://www.freedesktop.org/wiki/Specifications/shared-filemetadata-spec/

KDE uses user.xdg.tags and I think MATE also uses that.
https://www.reddit.com/r/linuxquestions/comments/sl7xfd/is_there_a_tag_based_file_manager/

Tags should be able to remain when moving the file, but if not cross-platform they should at least work between different linux DEs.

**Comment by Jeremy7701** [comment]
Emblems could work as tags, with the advantage of being very GUI-looking; they just need some bulk add/remove/list features adding.

**Reply by Qronikarz** [reply]
Kind of. 
They are GUI-looking, but some files don't show them. Like if the thumbnail is smaller than the emblem it doesn't render it. And that would be considered ok, but it also works that way on 1920x1080 wallpaper which is a norm nowadays. You need to zoom the icon size nearly up to max for it to show.

Also those 15 default emblems are not sufficient and it is not clear how to add new ones. The help button at the bottom is only pointing to https://www.linuxmint.com/documentation.php and I searched all 4 guides for "emblem" and not a single result came up. I know that it's the nemo extension and there are few instructions on its GitHub repository from other people, but casual users won't have idea where to look at.

For emblems to be considered as a tag they would need to allow custom names and to use them in search. And that's the bare minimum.

There's also for example: tag colors, tag file count, viewing all tags list, search combining like "boolean AND" + others.

I made a knowledge base about tags with 46 programs currently in it (Nemo included since it's a OS default browser) - https://github.com/Qronikarz/TagsResearch.

**Reply by Jeremy7701** [reply]
You can create your own (extra) ones.
They must be named `symbolic-xxxx.png` where xxxx is a custom name - I don't know if they need to be png's, but its a good idea if you need transparency.
They will appear in name order - I've put mine in a directory `~/.icons` - but I'm sure that other locations are probably better.
The following list probably duplicates some existing names from old or current choices:-

![Screenshot from 2024-04-28 22-15-15](https://github.com/linuxmint/discussions/assets/52585119/510d4b51-1f10-494c-9fc8-055b7ad4343c)

**Reply by Qronikarz** [reply]
> You can create your own (extra) ones.

I know that. That was not the point of my reply:
> I know that it's the nemo extension and there are few instructions on its GitHub repository from other people, but casual users won't have idea where to look at.

**Reply by MarjaE2** [reply]
In Finder, tags let me see file and folder status at a glance. They're colored dots next to the file names. I've had trouble with some tag colors being too similar, but otherwise they work well.

In Nemo, emblems don't. They're tiny icons superimposed on tiny icons.

I think 3 changes would help, until tags are ready:

1. allow users to add either colors or emblems to individual files, and
2. allow users to show either colors or emblems in their own column, not just superimposed on the folder icons.
3. add plain color emblems. What is "red"? What is "green"? - up to the user, but if you're using a color screen, it'll be clearly distinct.

**Comment by Jopp-gh** [comment]
Nemo needs tags, let me explain why.

Tags are like **quick shortcuts to folders** and an *update to bookmarks*  in Nemo. What does this mean? 

"Forget to dig manually in  paths and get where you want immediately! 
No more obsolete, manual browsing if not for exceptions"

Basically you'd use **bookmarked folders** as replacement to browse and in addition,  make most used (scored) folder names pop bolder add bigger in a tag cloud. 

### Tag panel
To make tags accessible, Nemo would need to  integrate a permanently visible panel at the bottom of each window like a film strip (horizontally) , bc words tend to expand horizontally and there is much more space in a horizontal panel. 
To balance the space taken of this tag cloud, the title and tool -bar of Nemo should merge too,  which is anyway over due,  imo.  

Our tag panel would be a **word cloud**, *maybe using some spare color to highlight important names* (eg. home folder dirs) and highlight most used paths *as folder names* with a moderate, larger font.
The advantage of a word cloud is that it allows to display much more (bookmarked) locations than the current implementation of bookmarks,  which is a kind of vertical bullet list (very space wasting and inefficient) in the file tree. 

### Tagging folders
Because, you bookmark (tag) folders not to tag folders but to move more quickly through your file system to find your files. 

### Custom folder tags
That's like having an alias - you can change an alias name to whatever you want and have the link nontheless pointing to the original location (folder). In addition, users should be able to define custom tags to group folders with tags, like in a search query. 

### Tagging files 
There should be yet another panel for file tags, instead to mix and mess folders and files together in one panel. why?  
In my opinion, "bookmarking" files is a bad idea. 
So you'd need custom names to group files in a meaningful way, even if files aren't in the same spot. Yeah, you could use the name of a folder as tag for files too but tags for files should be location independent. 

The "file tags" cloud and "folder tags" cloud could share the same space on the foot of Nemo's windows:  the folders cloud on the left half and the files cloud on the right half.

**Reply by Qronikarz** [reply]
> Tags are like **quick shortcuts to folders** and an _update to bookmarks_ in Nemo

Shortcuts maybe, but an update to bookmarks definitely no. Tags are tags, folders are folders and bookmarks are bookmarks.

> Basically you'd use **bookmarked folders** as replacement to browse

Not a replacement. Tags are used to aid the browsing when you want to find specific file. They don't require replacing anything.

> To make tags accessible, Nemo would need to integrate a permanently visible panel at the bottom of each window like a film strip (horizontally) , bc words tend to expand horizontally and there is much more space in a horizontal panel.

It can also be in already existing side panel as a tag list which is a concept which exists in many tagging programs already.

> To balance the space taken of this tag cloud, the title and tool -bar of Nemo should merge too, which is anyway over due, imo.

That I can agree with. Maybe it would be even possible to hide the "file, view, edit..." section and move the "grid/list view" buttons to the left so the "min,max,close" buttons can be moved in their place.

> The advantage of a word cloud is that it allows to display much more (bookmarked) locations than the current implementation of bookmarks, which is a kind of vertical bullet list (very space wasting and inefficient) in the file tree.

Vertical bullet list is how a majority of tagging programs do it - internet image boards, macOS Finder, KDE Dolphin, Spacedrive and many others.

I don't fully understand the last 3 sections of yours. 

> There should be yet another panel for file tags, instead to mix and mess folders and files together in one panel. why?

No. Ideally it wouldn't matter what is it that you're tagging. There's no need to divide it into folders and files.

**Comment by Jeremy7701** [comment]
You give files a file tag by putting them in a folder (or sub-folder).
Emblems would make ideal tags - if you can spare the programming effort (at a minimum) to allocate/remove multiple emblems in a single operation - somehow!

The screen layout ideas would be a separate issue.

**Reply by Qronikarz** [reply]
> You give files a file tag by putting them in a folder (or sub-folder).

No you do not. Folders are not tags and tags are not folders. There is a program that works that way - Hyperplane, but that is not the point of this discussion. 

> Emblems would make ideal tags

I would not say ideal, but they can be improved.

> if you can spare the programming effort (at a minimum) to allocate/remove multiple emblems in a single operation

That's not the only thing that is missing to transform emblems into tags

**Comment by Qronikarz** [comment]
Making an update here since it got some attention and want to make things clearer.

In the past I made a collection of notes and programs that are used to tag various things because I was developing my own tagging program. I made this into a knowledge database that is uploaded here: https://github.com/Qronikarz/TagsResearch

46 programs, most of which allow to tag your files, bookmarks, images or whatever else is left. And the file browsers included by default to see how many of them allow this feature.

When it comes to Linux there are 2 file browsers that already allow this feature and they both share the same tag specification - user.xdg.tags. Tags are stored in form of CSV in this field. However it comes with few problems - tags may not survive the file operations like copy or move if appropriate flags are not used. For more information there is the Extended Attributes page on ArchWiki.

If Mint's Nemo would also follow the user.xdg.tags way of storing tags then there would be 2 more operating systems that offer tags at default (Linux Mint Cinnamon and Solus Budgie). 

It could also be made into extension, but I think that it would be useful for users as a default option. nemo-emblems extension is using python so maybe this tool https://github.com/welbornprod/filetags can be used as a basis to make a new nemo-tags extension since it allows you to "easily edit the extended attributes, user.xdg.tags and user.xdg.comment, on files" and is also made with Python

Linux Mint in newest blog post wrote that they want to be more cross platform with "compatible software and common solutions". Following user.xdg.tags may be one of the things to do now and Nemo could join KDE's Dolphin and MATE's Caja with file tagging. (and possibly more Linux DEs to come later)

**Comment by Jeremy7701** [comment]
It's an impressive effort - but for a GUI file manager you need a GUI way of adding/removing tags in a safe way.
Presently, I find it irritating that I can't manipulate emblems over multiple files.
The actual information is easily accessible from the command line:-

```
jeremy@hal:~$ gio info -a metadata::emblems ~/PDF
uri: file:///home/jeremy/PDF
local path: /home/jeremy/PDF
unix mount: /dev/sdb4 /home ext4 rw,relatime,errors=remount-ro
attributes:
  metadata::emblems: [emblem-art]
```

**Reply by Qronikarz** [reply]
> but for a GUI file manager you need a GUI way of adding/removing tags in a safe way

That's a given. If we wanted CLI then I think you can do it even now with the FileTags tool I posted earlier. 

> Presently, I find it irritating that I can't manipulate emblems over multiple files.

Supposedly the emblems are stored in gvfs and it should be possible to write emblem to multiple files https://www.reddit.com/r/linuxmint/comments/z7r0tg/deleted_by_user/

However the emblems as they are, are only used in old GTK file browsers - Thunar, Mint, Caja. Two major desktops - GNOME and KDE don't use them. KDE allows you to set emblems for the tags - https://itsfoss.com/file-tagging-kde/#changing-emblems-of-tags

MATE Caja does both - emblems and tags. So they can coexist.

**Reply by Jeremy7701** [reply]
Mint have stated that they are sticking with GTK3 - is that "old GTK" - total rewrites aren't going to be possible.

**Comment by Jopp-gh** [comment]
@Qronikarz 
>Shortcuts maybe, but an update to bookmarks definitely no. Tags are tags, folders are folders and bookmarks are bookmarks.

Well,  what happens if you click on a bookmarked folder in the sidebar? 
Tags are a vehicle. 
Tags are "words" - but words do absolutely nothing by itself. 

User tags instead,  change/ evolve over time but still, 
 continue to point to files and folders.  

So tags are just meaningful if interwoven with files and folders. You don't tag to tag. You tag  for shortcuts to files and folders. It's all connected. 

You shouldn't be forced to tag everything by yourself, so my user friendly idea to "include" frequently visited folder names as "tags". 
In fact, the current conception "you've to tag everything",  makes  tagging so unattractive. 

>Not a replacement. Tags are used to aid the browsing when you want to find specific file. They don't require replacing anything

Maybe "replacement" sounds harsh - but I'm tired to navigate any file system (mac, linux,  win,  whatever) with a very old manual clickaround to get where I want.  Therefore I tried to explain a new way to  navigate a file system through folder names that are automatically added as "tags" if you visit those folders regularly.  Regularly visited folders would be added automatically to a "folder cloud". 

>Vertical bullet list is how a majority of tagging programs do it - internet image boards, macOS Finder, KDE Dolphin, Spacedrive and many others.

At this point I need to say, I don't try to emulate concepts seen in current Os's and I'm not interested in a "cool" look but a functional feature. 
My suggestion wouldn't cripple old and familiar mechanics but improve them and connect features instead. And I share this perspective in this thread. I didn't ask for help to correct my perspective and hope this doesn't sound harsh. 

Also the current standard is, to make no distinction between folders tags and file tags.  Mixing both files and folders causes more chaos instead of clarity,  at least from a programmers point of view.  

Basically I describe in my first post several tags: 
A) "folder tags" = bookmarked folders (to browse quicker) 
B)"folder tags" = grouped folders by an user defined word (to find quickly user defined folders)
C)"file tags" = grouped files by an user defined word (to find quickly user defined files)

meanwhile A would be a user friendly "integration with tags" and nothing more, B and C instead would be a new feature - to add,  with the help of "tags", user relevant files and folders.

**Reply by Qronikarz** [reply]
> Well, what happens if you click on a bookmarked folder in the sidebar?

It transports you to the folder that was bookmarked. So it's a kind of a shortcut to already existing place. 

> Tags are a vehicle.

I'd rather say they are identification method for describing and categorizing the items. They can then be used to search for the files with multiple tags instead of just one bookmark/category. So you can for example join `blue` and `sky` so it only shows you the files that have these 2 tags.

> You shouldn't be forced to tag everything by yourself

That's one of the downsides to the tagging for sure. But we currently live in AI boom era and object detection is really taking off. Even for text document you can use some LLM to quickly summarize it and assign few keywords. The only thing missing is file browsers offering the tag support so the programs that detect the objects have a place to write them to. There are already a few that do - File Explorer, Finder, Dolphin, Caja. But we're now talking about Nemo.

> so my user friendly idea to "include" frequently visited folder names as "tags". I'm tired to navigate any file system (mac, linux, win, whatever) with a very old manual clickaround to get where I want. Therefore I tried to explain a new way to navigate a file system through folder names that are automatically added as "tags" if you visit those folders regularly. Regularly visited folders would be added automatically to a "folder cloud".

That is an interesting concept that I could see replacing the current "recently visited/viewed" files or folders, but in no way it can be described as tagging system. You may be interested in looking up https://github.com/kra-mo/hyperplane which uses folder names to tag files, but that's the only thing from your idea and it is still in beta.

> I don't try to emulate concepts seen in current Os's and I'm not interested in a "cool" look but a functional feature.

Folders and files are true cross platform feature. Doesn't matter which OS - files and folders exist. Tagging is a little different. Windows offers tags, but only for a selected list of files. macOS offers great tagging support, and on Linux there are only 2 file browsers with tags. On both macOS and Linux it's possible to export the tags from files so they are not lost. So I would say that emulating concepts in current OSs is still important.

> Mixing both files and folders causes more chaos instead of clarity, at least from a programmers point of view.

Any examples? It works on macOS and maybe on Linux, but don't know - at least on Mint you can change the folder color so maybe it would be also possible to set extended attribute for the folder itself.

**Comment by Jopp-gh** [comment]
>I'd rather say they are identification method for describing and categorizing the items. 

Yes,  agreed! 
folders too are a way to categorize and organize - less flexible, but still a way to group common values. 
So my assertion, that folders can be seen as "weaker" tags

>They can then be used to search for the files with multiple tags instead of just one bookmark/category. So you can for example join `blue` and `sky` so it only shows you the files that have these 2 tags

Yeah, I imagine tag queries could be displayed like search windows,  as list or grid view. Multiple tags would show more or less output depending on the query 

>Folders and files are true cross platform feature. Doesn't matter which OS - files and folders exist. Tagging is a little different.

Exactly, so why not re-using and innovating semantic browsing instead to be happy to only assign "some abstract names"?

>macOS offers great tagging support,

As a ex-Mac user (for over 2 decades) I must say meh.. I had to use scripts to extend tags and make them useful  for everyday practice. By itself,  tags on mac are nice but not a finalized concept. 

>Any examples? It works on macOS and maybe on Linux, but don't know

Even before I started to use  pkm apps (2y ago or so) , tags were part of my routine to quality-control my files. Folders are just a convenience to organize files in a visual way on-disk. 
Nothing against folders, but in my opinion, there are just too much folders on current file systems - these hierarchies are ok for system folder organization but user folder organization should need much less folders and build instead,  on x-treme fast , semantic navigation, in addition to "stronger" tags to be defined by the user. So tagging gets a new Dimension.

**Comment by Jeremy7701** [comment]
If you haven't taken the trouble to organise your files into sensible folders and sub-folders etc then you definitely aren't going to bother organising a lot of files with tags.
I can't see it happening - unless it's some kind of automated process whereby files are analysed in some way to find "important" words that constitute likely tag words - in which case a search process can be just as effective.

**Reply by MarjaE2** [reply]
Some apps require their own libraries, and reorganizing files may break these libraries.

Sometimes it's also easier to organize a good copy and a bad copy of the same document in the same folder, with a tag to indicate that the good copy has also been imported into Calibre.

**Reply by Jachimo** [reply]
I don't think this is a good assumption. Mac OS X and its predecessors (possibly back to the IIgs deskop manager, I'm not sure) have had a limited form of file/folder tagging, called "Labels" for decades. And lots of people use it. I use it all the time when I'm working on a Mac, even for things as mundane as checking off which files in a directory I've looked at or reviewed for some purpose. Really handy.

More broadly, there are many cases where you need to organize files in two or three ways that are effectively orthogonal to each other. E.g. you might have a bunch of files organized into directories per date, but then you want to tag them by importance. The "importance" information is not captured in the file path / folder hierarchy. (You can kinda-sorta have one file in two places by using hard or soft links, but IMO it's pretty kludgy.)

The only good counterargument to implementing a feature like this at the filesystem (underlying metadata stored as extended attributes) and file manager (to display and edit) level, at least that I can come up with, is that for some types of files the information is more safely preserved by encoding it into the file itself using some file-format-specific metadata mechanism. (E.g. EXIF, ID3 tags, etc.) Because these are internal to the file itself, they tend to be more robust when files are moved around. This is a reasonable argument for e.g. music collection managers and the like.

Given that virtually all modern filesystems have the capability of storing extended attributes (or streams, or forks, or some similar construct) that arbitrary metadata could be stored in, it's really just the UI piece at the file-browser level that's missing. And that's really unfortunate IMO.

**Comment by Jopp-gh** [comment]
>> You shouldn't be forced to tag everything by yourself

>That's one of the downsides to the tagging for sure. But we currently live in AI boom era and object detection is really taking off

My decision and recent move to linux was  motivated by many reasons, the most  important reason was and is freedom,  understood as my right of privacy. People have rights too, not just duties. But rights aren't a casual gift. 
Since I depend on the success of my work, this means: my pc, my control. 

Artificial intelligence is outside my control and therefore not trustworthy. I hope Linux  stays away from getting sucked into this AI trend, bc let's be franc, who  decides what an Ai should do and not do based on inputs given?  Not even to talk what would happen if bots could get their own mind, the so called "singularity".  I'm consciously against AI. AI is no progress for humans but it's downfall and idiocy. If machines take over all our skills  humans become superfluous and I'm not sure if people realize this: 

"Sooner or later AI will dictate our behavior, through scores derived from our profiles if we dream just of benefits all times."

I trash everything that tries to toy with my peace,  no exception what Os I use. 
And if I need something that doesn't exist yet,  I can build my own scripts  and workflows. 
Because I'm rather new to linux, app development isn't yet on my list. So, right now,  I'm more than happy to contribute with lesser skills by sharing only my experience with the linux / Mint community.

**Comment by DIMFLIX** [comment]
@Qronikarz @Jopp-gh @MarjaE2 @Jeremy7701 @Jachimo  If the problem is still relevant for you, I have come up with a small solution. The [nemo-tags plugin](https://github.com/meowrch/nemo-tags) adds full tag support.

Unfortunately, I was unable to implement a drag-and-drop system or create a more attractive interface due to Nemo's limitations. But overall, I think it turned out to be quite usable and even convenient.

I would appreciate a star on GitHub repository as support.

<img width="1086" height="658" alt="preview" src="https://github.com/user-attachments/assets/fbf5ad8e-a991-4166-b72b-ae8cfa2b042b" />

**Reply by Qronikarz** [reply]
I don't use Mint since 22.1 beta, but it looks great.

Added it to my TagsResearch collection - https://github.com/Qronikarz/TagsResearch/blob/main/components/nemo-tags.md

**Reply by Jopp-gh** [reply]
Looks Like a Lovely Plugin! Thanks for sharing with us and for inviting us to enjoy your work!

Shortly after posting my comment (a year ago or so) I started to develop my own tag tool which I'm using till today. Still, next time when upgrading Mint, I'll take a look at your plugin too 😊 

A little critique, this plugin looks similar to classical tagging. That isn't wrong by any means, don't understand me wrong.
In my posts one year ago, I didn't just talk about classic tagging how everyone knows it, but about a system to navigate dynamically inside a tag cloud to reach folders. Since such a project requires more appropriate tools than bash and... if you are brave to build something unique and different, then let me know.

**Reply by DIMFLIX** [reply]
@Jopp-gh It would be interesting to take a look at your idea. Unfortunately, I don't know which specific posts you are referring to.

I am generally interested in developing this project, and you can leave a [feature request](https://github.com/meowrch/nemo-tags/issues) with a detailed description of your idea. 

If the functionality of the plugins in Nemo allows me to do so, then I may implement it (Nemo provides a limited API for interacting with its interface, which is why I was unable to implement tag viewing in the sidebar as I had previously wanted).

**Comment by Jeremy7701** [comment]
Pictorially, it looks just like a few emblems (but with smaller icons).

**Reply by DIMFLIX** [reply]
Yes, that's right. I simply don't know any other way to clearly display tags on files.

**Reply by Jeremy7701** [reply]
Emblems are stored in nemo in the metadata of each file - no need for a central JSON file. You can also get access via the CLI gio set/info utility.
The biggest problem is setting/changing file tags/emblems on a bulk scale. Setting files one at a time won't work.

**Comment by MarjaE2** [comment]
I had to switch to Windows; I tried to run Cinnamon in WSL2, since it's far better than the standard Windows desktop, but no luck.

That said, it might help to show the tag icons in different views, in light and dark mode, etc.

**Reply by DIMFLIX** [reply]
To be honest, I don't quite understand what you mean. The tag emblems are displayed quite well in both light and dark themes.
<img width="1052" height="648" alt="image" src="https://github.com/user-attachments/assets/a280c058-374c-478f-85d0-9a40e191c96e" />
